front_end/dart_tests/test_helpers.dart

```diff
@@ -0,0 +1,273 @@
+    }
+    return true;
+  }
+}
+
+/// A fake Http response we can give to JavaScript to mimic the server's compiled response.
+class FakeResponse {
+  text() => futureToPromise(Future.value(_text));
+  FakeResponse(this._text);
+  String _text;
+}
+
+/// An environment, with a prefix and a List of items, but no values.
+/// Fake version of what gets returned from Dart.environments
+class Environment {
+  String prefix;
+  List<String> items;
+  Environment(this.prefix, this.items);
+}
+
+/// Create a new execution context with hard-coded fields as if we were
+/// executing a file from the web directory, so it can't just be imported with a
+/// package: URL.
+newContextInWeb() {
+  var context = FakeExecutionContext();
+  context.selectedCallFrame = allowInterop(() => FakeCallFrame.inWeb()
+    ..evaluationResult = FakeRemoteObject(value: 'foo/bar/web/main.html'));
+  return context;
+}
+
+/// Create a new execution context with hard-coded fields as if we were
+/// executing a file from the lib directory, so normal importing works.
+newContextInLib() {
+  var context = FakeExecutionContext();
+  context.selectedCallFrame = allowInterop(() => FakeCallFrame.inLib()
+    ..evaluationResult = FakeRemoteObject(value: 'foo/bar/web/main.html'));
+  return context;
+}
+
+/// A fake execution context. We write the fake in JS and access it using JS
+/// interop because we access JS methods that start with underscores, and if we
+/// create it in Dart those will get mangled and be harder to access.
+@JS('Dart.FakeExecutionContext')
+class FakeExecutionContext {
+  @JS('selectedCallFrame')
+  external set selectedCallFrame(x);
+}
+
+/// A fake that we use to replace the devtools Bindings namespace.
+class FakeChromeBindings {
+  get debuggerWorkspaceBinding => this;
+  UILocation _location = UILocation();
+  rawLocationToUILocation(_) => _location;
+  FakeChromeBindings({String currentFileUrl}) {
+    _location._url = currentFileUrl;
+  }
+  set currentFileUrl(x) => _location._url = x;
+}
+
+/// A fake for devtools UILocation.
+///
+/// Some of the things we provide really belong on other objects,
+/// but we add them here to avoid having quite so many fakes.
+class UILocation {
+  get columnNumber => 7;
+  get lineNumber => 42;
+
+  /// Just use this as its own uiSourceCode.
+  get uiSourceCode => this;
+  // url would really be on UISourceCode.
+  var _url;
+  url() => _url;
+}
+
+/// A fake for the devtools SDK.DebuggerModel.CallFrame.
+class FakeCallFrame {
+  functionLocation() => this;
+  script() => this;
+  get sourceURL => 'someURL';
+
+  /// If true we retain scopeChainForWeb, otherwise scopeChainForLib.
+  bool inWeb;
+  scopeChain() => inWeb ? scopeChainWeb() : scopeChainLib();
+
+  /// Allow creating a call frame in one of two flavors, one hard-coded to a
+  /// particular file in the web directory, and another in lib.
+  FakeCallFrame(this.inWeb);
+  FakeCallFrame.inWeb() : this.inWeb = true;
```

Comment:
remove extra this. You have been writing too much JS :)

Reply:
Done.